### PR TITLE
Contribute.md Guide for JS Examples

### PR DESCRIPTION
**Pull request by maboa**
This is the first pass of the contributing guide. I imagine it will require a few iterations of reviewing and modification :)

### CODE REVIEW

File: CONTRIBUTE.md [hunk 1]
```diff
@@ -0,0 +1,333 @@
+# JavaScript Interactive Examples Contributing Guide
+
+A guide for people wanting to contribute interactive JavaScript examples.
+
+## Naming Things
```
Comment (wbamberg) [inline CONTRIBUTE.md]: "Naming Variables" might be better.

File: CONTRIBUTE.md [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+console.log(array1);
+// expected output: Array ["a", "b", "c"]
+
+var array2 = [1, 2, 3];
+
+console.log(array2);
+// expected output: Array [1, 2, 3]
+```
+
+An alternative to using numerical suffixes could be to use more descriptive suffixes, such as `arrayLetters` and `arrayNumbers` in the above examples.
```
Comment (wbamberg) [inline CONTRIBUTE.md]: Should this be in a separate section to "Numerical Suffixes"? Like "Descriptive suffixes" or something?

File: CONTRIBUTE.md [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+This may make more sense in the context of the following example:
+
+```
+var collatorDe = new Intl.Collator('de', { usage: 'search', sensitivity: 'base' });
+var collatorFr = new Intl.Collator('fr', { usage: 'search', sensitivity: 'base' });
+```
+
```
Comment (wbamberg) [inline CONTRIBUTE.md]: Is it worth having a third section like "descriptive names". For example:

```
let fruit = ["apples", "bananas", "oranges"]
```

...this only works for less abstract examples. But less abstract example might be better anyway, as the "tangible examples" section says.

File: CONTRIBUTE.md [hunk 4]
```diff
@@ -0,0 +1,333 @@
+# JavaScript Interactive Examples Contributing Guide
```
Comment (wbamberg) [inline CONTRIBUTE.md]: I think "Style Guide" would be better than "Contributing Guide". There's a lot else to contributing than the style of the examples, which is all that this guide covers.

File: CONTRIBUTE.md [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+
+Occasionally browsers will produce different results, in these cases we can denote the varying output like so:
+
+```
+console.log(matchesDe);
+// expected output (Chrome / Firefox): Array ["Bären"]
+// expected output (Safari): Array ["Bären", "Baren"]
+```
+
+## Language Choice
```
Comment (wbamberg) [inline CONTRIBUTE.md]: Could call this "Language Choice (ES5 versus ES6)" just to be clear that this s the choice.

File: CONTRIBUTE.md [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+ES6 examples should use:
+
+- `let` and `const` instead of `var`
+- Arrow functions (`=>`) for Anonymous Functions (see below)
+- Template literals (``string text ${expression} string text``)
+- Spread syntax (`myFunction(...iterableObj);`)
+
+## Providing Context
+
+Where possible it may be useful to provide examples showing how a method could be used in the context of a function or class. It may not always be possible to provide a succinct example and a series of `console.log`s can also be a valid way of illustrating an example.
```
Comment (wbamberg) [inline CONTRIBUTE.md]: An example would be good here.

File: CONTRIBUTE.md [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+
+## Tangible examples
+
+It's nice to be able to illustrate a programming concept using real or imaginary world examples, if possible using the same theme throughout a specific topic.
+
+For example:
+
+```
+function 1(disposition) {
```
Comment (wbamberg) [inline CONTRIBUTE.md]: Is "1" here and below intentional, or do you mean "monster1"?

File: CONTRIBUTE.md [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+} catch (err) {
+  console.log(err);
+  // expected output: RangeError: invalid language tag: EN_US
+}
+```
+
+## Format
+
+The aim is to try and fit examples into 12 lines of code, if it's not possible to create a useful example in 12 lines, it's fine to use more lines. We currently have examples of up to 25 lines.
+
```
Comment (wbamberg) [inline CONTRIBUTE.md]: I would move this section out of here and into its own subsection, at the top, called "Example size" or something. For me this is not a coding style thing exactly, and it is important enough to highlight.

Comment (wbamberg) [reply CONTRIBUTE.md]: I'd also make the language a bit stronger here. It's always tempting to go above 12, but we really really don't want to unless we have to.

File: CONTRIBUTE.md [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+```
+try {
+  Intl.getCanonicalLocales('EN_US');
+} catch (err) {
+  console.log(err);
+  // expected output: RangeError: invalid language tag: EN_US
+}
+```
+
+## Format
```
Comment (wbamberg) [inline CONTRIBUTE.md]: I would call this "JavaScript coding style"

File: CONTRIBUTE.md [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+An alternative to using numerical suffixes could be to use more descriptive suffixes, such as `arrayLetters` and `arrayNumbers` in the above examples.
+
+This may make more sense in the context of the following example:
+
+```
+var collatorDe = new Intl.Collator('de', { usage: 'search', sensitivity: 'base' });
+var collatorFr = new Intl.Collator('fr', { usage: 'search', sensitivity: 'base' });
+```
+
+### Indicating console.log output
```
Comment (wbamberg) [inline CONTRIBUTE.md]: I'm not sure if this is intended to be an H3 under "Naming Things"

File: CONTRIBUTE.md [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+
+To indicate the ouput we expect, we place a comment `// expected output:` on the line below each `console.log`.
+
+
+### Representing Browser Differences
```
Comment (wbamberg) [inline CONTRIBUTE.md]: Again I'm not sure if this is intended to be an H3 under "Naming Things"

File: CONTRIBUTE.md [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+- Template literals (``string text ${expression} string text``)
+- Spread syntax (`myFunction(...iterableObj);`)
+
+## Providing Context
+
+Where possible it may be useful to provide examples showing how a method could be used in the context of a function or class. It may not always be possible to provide a succinct example and a series of `console.log`s can also be a valid way of illustrating an example.
+
+## Tangible examples
+
+It's nice to be able to illustrate a programming concept using real or imaginary world examples, if possible using the same theme throughout a specific topic.
```
Comment (schalkneethling) [inline CONTRIBUTE.md]: Nit: "When illustrating a programming concept, it is beneficial to mimic real world examples as much as possible. Once you have chosen your real world theme, stick to the theme throughout the specific example."

File: CONTRIBUTE.md [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+var proxy1 = new Proxy(1, handler1);
+
+console.log(new proxy1("fierce").disposition);
+// expected output: "fierce"
+```
+
+
+## Dealing with Errors
+
+Writing code that will throw an error can be useful way to illustrate an example. However an uncaught error will halt execution immediately and will not display the results of any `console.log` statements in the example.
```
Comment (schalkneethling) [inline CONTRIBUTE.md]: Nit: "throw an error can be *a* useful way to" "However[,] an uncaught error"

File: CONTRIBUTE.md [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+
+If you wish to use an error to illustrate a method you could wrap it in a `try` `catch` clause, for example:
```
Comment (schalkneethling) [inline CONTRIBUTE.md]: Nit: "If you wish to use an error to illustrate a method, wrap it in a `try/catch` block for example:"

File: CONTRIBUTE.md [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+
+```
+var array1 = [2, 5, 7, 9];
+```
+
+Note the spacing after the commas.
+
+### Passing parameters
+
+We space after commas when passing parameters to a function, but no space padding of parenthesis:
```
Comment (schalkneethling) [inline CONTRIBUTE.md]: Nit: "We use a space after commas when passing parameters to a function, but do not pad the parenthesis: i.e.
```
calcAngle(8, 10);
```
but not,

```
calcAngle( 8, 10 );
```
"

File: CONTRIBUTE.md [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+
+Generally functions are defined with the `function` keyword at the beginning:
+
+```
+function sum(a, b) {
+  return a + b;
+}
+```
+
+Note the space before the opening curly bracket.
```
Comment (schalkneethling) [inline CONTRIBUTE.md]: Nit: "bracket" => "brace"

File: CONTRIBUTE.md [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+function sum(a, b) {
+  return a + b;
+}
+```
+
+Note the function name starts with a lowercase letter and can be camelCased.
+
+### Class Definition
+
+As classes came in with ES6 we can use other ES6 concepts when defining them such as `const`. We also need to be aware of the changing scope of `this`.
```
Comment (schalkneethling) [inline CONTRIBUTE.md]: Nit: "As classes were introduced with ES6, we"

File: CONTRIBUTE.md (hunk 4 above)
Comment (maboa) [reply CONTRIBUTE.md]: :+1:

File: CONTRIBUTE.md (hunk 1 above)
Comment (maboa) [reply CONTRIBUTE.md]: 👍

File: CONTRIBUTE.md (hunk 2 above)
Comment (maboa) [reply CONTRIBUTE.md]: 👍

File: CONTRIBUTE.md (hunk 3 above)
Comment (maboa) [reply CONTRIBUTE.md]: Yes we could do - I'm torn about whether to be consistent or allowing the flexibility to be descriptive - and I'm fine with both, but thinking now I may have been leaning to strongly on the consistency side and so we may want to change back some naming of the array examples I recently PR'd (https://github.com/mdn/interactive-examples/pull/350)

File: CONTRIBUTE.md (hunk 10 above)
Comment (maboa) [reply CONTRIBUTE.md]: I think it works under "Naming Things" ... kinda - as we're talking about the naming/wording of a comment. Loose I know :) But in view of the fact that we're changing "Naming Things" to "Naming Variables" then I think we can reorganise a bit.

File: CONTRIBUTE.md (hunk 11 above)
Comment (maboa) [reply CONTRIBUTE.md]: Similar to above, it's more wording than naming.

File: CONTRIBUTE.md (hunk 5 above)
Comment (maboa) [reply CONTRIBUTE.md]: Sure.

File: CONTRIBUTE.md (hunk 6 above)
Comment (maboa) [reply CONTRIBUTE.md]: 👍

File: CONTRIBUTE.md (hunk 7 above)
Comment (maboa) [reply CONTRIBUTE.md]: Yup - the monster part got eaten somehow.

File: CONTRIBUTE.md (hunk 9 above)
Comment (maboa) [reply CONTRIBUTE.md]: 👍

File: CONTRIBUTE.md (hunk 8 above)
Comment (maboa) [reply CONTRIBUTE.md]: OK - agree!

File: CONTRIBUTE.md [hunk 18]
```diff
@@ -0,0 +1,333 @@
+# JavaScript Interactive Examples Contributing Guide
+
+A guide for people wanting to contribute interactive JavaScript examples.
+
+## Naming Things
+
+It probably makes sense to name things in a consistent way throughout all examples.  Where possible the naming should ‘reflect’ the type of that object or variable.
```
[the remark this answers is not in the document]
Comment (maboa) [reply CONTRIBUTE.md]: No worries Schalk. Thanks for taking the time to feedback.

Agree with your comment :)

File: CONTRIBUTE.md [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+```
+object1 = {};
+var string1 = “a string”;
+var array1 = ['a', 'b', 'c'];
+var array2 = [1, 2, 3];
+```
+
+Note it is suggested that even if there is only one instance of that object then it should be suffixed with a `1`.
+
+Note also although we could use the same name for multiple instances, it's suggested that we use a progressive numbering system to minimise confusion should the reader want to experiment with the example code.
```
[the remark this answers is not in the document]
Comment (maboa) [reply CONTRIBUTE.md]: 👍

File: CONTRIBUTE.md [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+
+```
+console.log(matchesDe);
+// expected output (Chrome / Firefox): Array ["Bären"]
+// expected output (Safari): Array ["Bären", "Baren"]
+```
+
+## Language Choice
+
+For more established example content (such as arrays) it is recommended that we stick to ES5, where examples are required for APIs standardised after ES6 it is felt that we can use ES6 to illustrate the examples.
```
[the remark this answers is not in the document]
Comment (maboa) [reply CONTRIBUTE.md]: 👍

File: CONTRIBUTE.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (maboa) [reply CONTRIBUTE.md]: 👍 Presuming you meant `In instances ...`

File: CONTRIBUTE.md (hunk 12 above)
Comment (maboa) [reply CONTRIBUTE.md]: 👍

File: CONTRIBUTE.md [hunk 21]
```diff
@@ -0,0 +1,333 @@
...
+For example:
+
+```
+function 1(disposition) {
+  this.disposition = disposition;
+}
+
+var handler1 = {
+  construct: function(target, args) {
+    console.log("1 constructor called");
```
[the remark this answers is not in the document]
Comment (maboa) [reply CONTRIBUTE.md]: 👍

File: CONTRIBUTE.md [hunk 22]
```diff
@@ -0,0 +1,333 @@
...
+    // expected output: " constructor called"
+
+    return new target(...args);
+  }
+};
+
+var proxy1 = new Proxy(1, handler1);
+
+console.log(new proxy1("fierce").disposition);
```
[the remark this answers is not in the document]
Comment (maboa) [reply CONTRIBUTE.md]: 👍

File: CONTRIBUTE.md (hunk 13 above)
Comment (maboa) [reply CONTRIBUTE.md]: 👍

File: CONTRIBUTE.md (hunk 14 above)
Comment (maboa) [reply CONTRIBUTE.md]: OK - I'm trying not to be too prescriptive, but maybe the intention of this doc is to be prescriptive. So 👍 :)

File: CONTRIBUTE.md (hunk 15 above)
Comment (maboa) [reply CONTRIBUTE.md]: 👍

File: CONTRIBUTE.md (hunk 16 above)
Comment (maboa) [reply CONTRIBUTE.md]: 👍

File: CONTRIBUTE.md (hunk 17 above)
Comment (maboa) [reply CONTRIBUTE.md]: 👍

**Comment by maboa** [comment]
OK - I will take all your comments on board and create a new revision of the doc. Thanks!

File: CONTRIBUTE.md (hunk 14 above)
Comment (maboa) [reply CONTRIBUTE.md]: Also I just noticed you changed the 'phrasing' of `try/catch` which is fine.

File: CONTRIBUTE.md [hunk 23]
```diff
@@ -0,0 +1,333 @@
...
+
+We chose clarity over brevity when defining object properties, for example:
+
+```
+var object1 = {
+  property1: 42,
+  property2: 'foo'
+};
+```
+
```
Comment (chrisdavidmills) [inline CONTRIBUTE.md]: For this one, maybe give an example of what not to do?

File: CONTRIBUTE.md [hunk 24]
```diff
@@ -0,0 +1,333 @@
...
+```
+object1 = {};
+var string1 = “a string”;
+var array1 = ['a', 'b', 'c'];
+var array2 = [1, 2, 3];
+```
+
+Note it is suggested that even if there is only one instance of that object then it should be suffixed with a `1`.
+
```
Comment (chrisdavidmills) [inline CONTRIBUTE.md]: I find language like this a bit too wishy washy. I think you should be more prescriptive and direct, and ummm, less English ;-)

So for example

"Note that even if there is only one instance of that object, it should be suffixed with a `1`."

Comment (maboa) [reply CONTRIBUTE.md]: Right-ho! ;)

File: CONTRIBUTE.md (hunk 23 above)
Comment (maboa) [reply CONTRIBUTE.md]: Done.

**Pull request merged**